Comments

jonahragle opened the pull request.

jonahragle committed:
s

jonahragle committed:
saving my work

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        String authorId = sessionManager.getCurrentUserId();
+        String postId = paramHelper.valueOf("postId");
+        if((bodyText.isEmpty() || bodyText == null) && (authorId.isEmpty() || authorId == null) && (postId.isEmpty() || postId == null)){
+            return "bad";
```
should return a JSON string. also instead of returning, use the halt() function provided by Spark.

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        String bodyText = paramHelper.valueOf("bodyText");
+        String authorId = sessionManager.getCurrentUserId();
+        String postId = paramHelper.valueOf("postId");
+        if((bodyText.isEmpty() || bodyText == null) && (authorId.isEmpty() || authorId == null) && (postId.isEmpty() || postId == null)){
```
just check for null, not empty. RequestParamHelper will do that for you. this might be why the the server crashes when any of these are missing from the request.

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        commentModel.setBodyText(bodyText);
+        commentModel.setDate();
+        mongoCollection.save(commentModel);
+        return postId;
```
should return a JSON string

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        commentModel.setPostId(postId);
+        commentModel.setAuthorId(authorId);
+        commentModel.setBodyText(bodyText);
+        commentModel.setDate();
```
setters should take in parameters

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        mongoCollection.save(commentModel);
+        return postId;
+    });
+
```
### `POST /comments/create`

- doesn't return JSON. Using an instance of the GeneralResponse class might help with this.
- server crashes with 500 error when `bodyText` is empty
  - server crashes when empty request body sent
  - should halt with an error instead
- still executes when a postId is not provided - this should halt with an error

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        RequestParamHelper paramHelper = new RequestParamHelper(request);
+        String postId = paramHelper.valueOf("postId");
+        String startAfterId = paramHelper.valueOf("startAfterId");
+        String date = paramHelper.valueOf("date");
```
unsure why date is required for listing comments, unless fetching comments from a specific date range?

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        String postId = paramHelper.valueOf("postId");
+        String startAfterId = paramHelper.valueOf("startAfterId");
+        String date = paramHelper.valueOf("date");
+        PostModel post = mongoCollectionPost.findOne(eq("_id", new ObjectId(postId)));
```
postId was not null checked before usage - this could be why the server crashes with a 500 error

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        PostModel post = mongoCollectionPost.findOne(eq("_id", new ObjectId(postId)));
+        List<CommentModel> commentModelList;
+        if(post == null){
+            return "post DNE";
```
should halt with a JSON string

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+                DateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
+                date = dateFormat.format(thedate);
+            }
+            commentModelList = getByDate(postId, date);
```
this will cause comments from the current day to be fetched. the spec requires that if `startAfterId` is missing, to return an initial list of ALL comments for the post, not just for the specific day

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        }else{
+            commentModelList = getAfterId(postId, startAfterId);
+        }
+        if(postId == null){
```
this null check is performed AFTER postId has been used.

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        return commentList;
+
+    });
+
```
### `GET /comments/list`
- `startAfterId` either broken or does not work as intended
- `limit` parameter is missing from endpoint
- `date` parameter not required from spec, but:
  - current implementation will only list comments from the current DAY when `startAfterId` is not supplied, not all comments for the specific post
- blank postId crashes server with 500 error, along with a blank request
  - should halt with an error instead

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        if(post == null){
+            return commentModels;
+        }else{
+            iterable = mongoCollectionComments.find(Filters.and(Filters.eq("post_id", postId), Filters.gt("_id", startAfterId.toString())));
```
this actually doesn't have to be in an else block - since IF post == null, the function is returned and the remaining lines are not executed. 

- also, we are not trying to compare postId's, we are trying to compare commentId's (since we want to page the list of comments, not the list of posts). 
- startAfterId.toString() is not required since startAfterId is already a String. also, "_id" isn't a String, but an ObjectId, so you will probably want to pass `new ObjectId(startAfterId)`.
   - this is most likely the culprit as to why `startAfterId` in `/comments/list` is broken

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        iterable.into(commentList);
+        return commentList;
+    }
+
```
fetching by date is not required. generally this method is pretty good. Only two things I could say is use a variable for a fetch limit (that way the limit can easily be changed or controlled). Also, don't pass a string for the date to the database, but an actual date object. this is absolutely not a requirement, since passing a String will definitely work, but passing a date object causes it to be converted into seconds, which is faster for the database to compare (will cause this route to perform faster).

other than that, this function is great.

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        RequestParamHelper paramHelper = new RequestParamHelper(request);
+        CommentModel commentModel;
+        //PermissionChecker permissionChecker;
+        String commentId = paramHelper.valueOf("_id");
```
this should probably be `paramHelper.valueOf("comment_id")` instead of `paramHelper.valueOf("_id")` so that it is uniform with the rest of the API

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        //PermissionChecker permissionChecker;
+        String commentId = paramHelper.valueOf("_id");
+        String bodyText = paramHelper.valueOf("bodyText");
+        commentModel = mongoCollection.findOne(eq("_id", new ObjectId(commentId)));
```
commentId is not null checked, this might be why the server crashes with a 500 error

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        if(sessionManager.getCurrentUserId().equals(commentModel.authorId)){
+            commentModel.setBodyText(bodyText);
+            mongoCollection.findOneAndReplace(eq("_id", new ObjectId(commentId)), commentModel);
+            return " ok ";
```
should return JSON

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+            mongoCollection.findOneAndReplace(eq("_id", new ObjectId(commentId)), commentModel);
+            return " ok ";
+        }
+        return " not ok ";
```
should halt with JSON

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        }
+        return " not ok ";
+    });
+
```
### `PUT /comment/update`

- server crashes with 500 error when request body is blank - should error out
  - server crashes with 500 error when request has missing `_id` parameter
  - should halt with an error instead
- still executes when body text is missing
  - should halt with an error instead
- not too big of a deal, but `_id` should probably be something like `comment_id` so that the REST api is uniform

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+/*        UserModel userModel = new UserModel();
+        userModel.setAdmin(false);*/
+        String commentId = paramHelper.valueOf("_id");
+        CommentModel commentModel = mongoCollection.findOne(eq("_id", new ObjectId(commentId)));
```
`commentId` is not null checked before usage

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        String commentId = paramHelper.valueOf("_id");
+        CommentModel commentModel = mongoCollection.findOne(eq("_id", new ObjectId(commentId)));
+        if(commentModel == null){
+            return "comment DNE";
```
should halt with JSON

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        if(commentModel == null){
+            return "comment DNE";
+        }
+        if(userModel.isAdmin()){
```
this only allows admins to delete comments - the idea is that admins, moderators, users can delete comments.

do not worry about permission checking for now - allow any user to delete/update a comment. Permission checking for whom can alter a comment is pretty tricky and I will add the permission checks in.

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        }
+        if(userModel.isAdmin()){
+            result = mongoCollection.deleteOne(eq("_id", new ObjectId(commentId)));
+            return " ok ";
```
should return JSON

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+            result = mongoCollection.deleteOne(eq("_id", new ObjectId(commentId)));
+            return " ok ";
+        }
+        return " not authorized ";
```
should return JSON

chuzarski commented on src/main/java/com/umbr3114/controllers/PostController.java:
```diff
@@ -50,6 +50,10 @@
+        postModel = new PostModel();
+        postModel.setAuthorId(authorId);
+        postModel.setBodyText(bodyText);
+        postModel.setTitle(title);
```
lines 53-56 were added for some reason and this might be an accident from cleaning up a merge conflict. just a heads up.

chuzarski commented on src/main/java/com/umbr3114/models/CommentModel.java:
```diff
@@ -0,0 +1,59 @@
+
+    public String getDateString() { return dateString; }
+
+    public void setDate() { this.date = new java.util.Date(System.currentTimeMillis());
```
this setter should not be creating the date, but taking in a date as a parameter. since date is a public variable in the ModelClass, a setter is also not required, since anybody has access to the variable.

chuzarski commented on src/main/java/com/umbr3114/models/PostModel.java:
```diff
@@ -16,6 +20,7 @@
     public String dropId;
     public String idString;
     //public List<CommentModel> comments;
+    public int startAfter;
```
startAfter should not be in the PostModel. Posts do not have to contain data about their position in a list. Also, id's aren't stored into `int` variables.

chuzarski commented on src/main/java/com/umbr3114/models/PostModel.java:
```diff
@@ -50,9 +55,14 @@ public void setAuthorId(String authorId) {
 
+    //public void addComment(CommentModel comment){
+    //    comments.add(comment);
+    //}
```
commented out code should be cleaned up

chuzarski commented on src/main/java/com/umbr3114/controllers/CommentController.java:
```diff
@@ -0,0 +1,160 @@
+        }
+        return " not authorized ";
+    });
+}
```
### `DELETE /comment/destroy`

- 500 server error when request is sent with blank body 
  - should halt and throw error instead
- only admins are able to delete comments (authors, moderators and admins should be able to delete)
  - (see overall notes, do not worry about permission checking right now)

chuzarski reviewed (changes requested):
### `POST /comments/create`

- doesn't return JSON. Using an instance of the GeneralResponse class might help with this.
- server crashes with 500 error when `bodyText` is empty
  - server crashes when empty request body sent
  - should halt with an error instead
- still executes when a postId is not provided - this should error out

### `GET /comments/list`
- `startAfterId` either broken or does not work as intended
- `limit` parameter is missing from endpoint
- `date` parameter not required from spec, but:
  - current implementation will only list comments from the current DAY when `startAfterId` is not supplied, not all comments for the specific post
- blank `postId` crashes server with 500 error, along with a blank request
  - should halt with an error instead
- PostModel should not require a `startAfter` variable.

### `PUT /comment/update`

- server crashes with 500 error when request body is blank - should error out
  - server crashes with 500 error when request has missing `_id` parameter
  - should halt with an error instead
- still executes when body text is missing
  - should halt with an error instead
- not too big of a deal, but `_id` should probably be something like `comment_id` so that the REST api is uniform

### `DELETE /comment/destroy`

- 500 server error when request is sent with blank body 
  - should halt and throw error instead
- only admins are able to delete comments (authors, moderators and admins should be able to delete)
  - (see overall notes, do not worry about permission checking right now)

## overall
- missing documentation on how to use endpoints (docs/Endpoints.md needs to be updated and/or the entry in the Github wiki)
- do not worry about permission checks, i'll write the permission check handler (it is tricky for the case of comments)
- most endpoints do not match spec (do not return JSON, parameters are unpredictable)
  - look at usage of GeneralResponse class in the PostController
- checking for empty values needs to be improved
  - server crashes with a 500 error in every endpoint

jonahragle committed:
comments

chuzarski committed:
fixed merge conflict - merging master into comments

chuzarski committed:
Change log:

Merged latest master into branch

> Main.java

- Added Auth checks to the following routes:
  - `/comments/create`
  - `/comment/update`
  - `/comment/destroy`

> CommentController.java

- Added `LIMIT_DEFAULT` constant

> saveComments

- fixed authorId so that it is pulled from sessionManager (the original implementation was correct since the spec was out of date and the project has advanced since then)
- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST` when malformed JSON sent in

> updateComments

- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST` when `comment_id` and `bodyText` missing
- Added a null check for commentModel when it is fetched from database (a comment that is requested that does not exist would crash the Route)
  - Added 404 response
- Added permission checks to Route

> destroyComments

- Adjusted GeneralResponse error to return HTTP code `400 BAD REQUEST` when param data is missing
- Changed input parameter `_id` to `comment_id` to keep route predictable
- Adjusted commentModel null check to halt and throw a 404 error
- removed DeleteResult variable (since fetching the comment from the database is successful by the point the delete is performed)
- added permission checks to Route

> getComments

- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST`
- Reformatted function for readability (variables declared at top, assigned in body)
- changed `lim` variable to default to `LIMIT_DEFAULT` class constant
- Added nullcheck to limit parameter
- Removed check for post model (unneccsary in comment loader). If the post does not exist, a blank list of comments will be returned

> getAfterId

- Removed usage of PostModel (not sure why it was in here. Check exsistence of the post? That is already implied since the comment was created in the first place, thus that check was already done)
- null check on startAfterId (the spec calls that startAfterId is optional)
  - a default list is now fetched based when startAfterId is not present

chuzarski commented:
I updated this branch to master and also added the permission checks like I said I would write.
There were still a few issues with this branch, found from testing. 
This P/R is now good to pull in

Here is a total change log of all changes I made:

Merged latest master into branch
Documentation for endpoints written

> Main.java

- Added Auth checks to the following routes:
  - `/comments/create`
  - `/comment/update`
  - `/comment/destroy`

> CommentController.java

- Added `LIMIT_DEFAULT` constant


> saveComments

- fixed authorId so that it is pulled from sessionManager (the original implementation was correct since the spec was out of date and the project has advanced since then)
- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST` when malformed JSON sent in

> updateComments

- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST` when `comment_id` and `bodyText` missing
- Added a null check for commentModel when it is fetched from database (a comment that is requested that does not exist would crash the Route)
  - Added 404 response
- Added permission checks to Route

> destroyComments

- Adjusted GeneralResponse error to return HTTP code `400 BAD REQUEST` when param data is missing
- Changed input parameter `_id` to `comment_id` to keep route predictable
- Adjusted commentModel null check to halt and throw a 404 error
- removed DeleteResult variable (since fetching the comment from the database is successful by the point the delete is performed)
- added permission checks to Route

> getComments

- Adjusted GeneralResponse errors to return HTTP code to `400 BAD REQUEST`
- Reformatted function for readability (variables declared at top, assigned in body)
- changed `lim` variable to default to `LIMIT_DEFAULT` class constant
- Added nullcheck to limit parameter
- Removed check for post model (unneccsary in comment loader). If the post does not exist, a blank list of comments will be returned

> getAfterId

- Removed usage of PostModel (not sure why it was in here. Check exsistence of the post? That is already implied since the comment was created in the first place, thus that check was already done)
- null check on startAfterId (the spec calls that startAfterId is optional)
  - a default list is now fetched based when startAfterId is not present
